### PR TITLE
Adding a well-known endpoint for public key of mock uss

### PR DESCRIPTION
**Pull request by punamverma**
For FAA UFT message signing needs to be done between uss-uss communication. The public key needs to be provided at  a well-known endpoint. This PR implements the endpoint. Mock keys will be added to the keys folder. Waiting for mock keys from FAA, as the USSes need to validate the keys using SCVP validation tool. Or else, we will add both mock keys, and add a requirement for USSes to disable SCVP validation when running interuss tests.

### CODE REVIEW

File: monitoring/messagesigning/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# start by pulling the python image
```
Comment (barroco) [inline monitoring/messagesigning/Dockerfile:1]: By moving your code to the mock_uss module, this file will not be required anymore. 
Instead, create a `run_locally_[SERVICE_NAME].sh` similar to `[run_locally_scdsc.sh](https://github.com/interuss/dss/blob/master/monitoring/mock_uss/run_locally_scdsc.sh)`.

Based on the design document requested in #893, we will be able to agree on the actual SERVICE_NAME and if we need a new script or should update the existing one.

For your information, the docker image which will be used is defined in [`monitoring/Dockerfile`](https://github.com/interuss/dss/blob/master/monitoring/Dockerfile)
Please do not modify the docker image itself. 
If you need to use the keys defined in the configuration below, please use a volume mount in the shell script like [this](https://github.com/interuss/dss/blob/master/monitoring/mock_uss/run_locally_scdsc.sh#L30)

File: monitoring/messagesigning/run_well_known.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
Comment (barroco) [inline monitoring/messagesigning/run_well_known.sh:1]: This file should not be required anymore since it will be replaced by the new `mock_uss/run_locally_[SERVICE_NAME].sh`.

File: monitoring/messagesigning/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+flask==1.1.2
```
Comment (barroco) [inline monitoring/messagesigning/requirements.txt:1]: This file should not be required anymore since dependencies are provided by the [monitorlib requirements](https://github.com/interuss/dss/blob/master/monitoring/monitorlib/requirements.txt).

File: monitoring/messagesigning/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+loguru
```
Comment (barroco) [inline monitoring/messagesigning/requirements.txt:2]: I see in your other PRs that you used extensively this library, is it a requirement for the feature you are contributing or can we live without it ?
We are making our best to limit the number of dependencies to limit the security risks and size of our docker images. New additions must be motivated.

File: monitoring/messagesigning/config.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import os
+
+
+class Config:
+    BASE_PATH = os.path.dirname(__file__)
+    PRIVATE_KEY_PATH = "{}/keys/mock_priv.pem".format(BASE_PATH)
```
Comment (barroco) [inline monitoring/messagesigning/config.py:6]: Those configuration input should be moved to `mock_uss/config.py`. Again, depending on the design document requested in https://github.com/interuss/dss/pull/893, we can agree on the name of those configuration flags.

File: monitoring/messagesigning/mockuss_well_known_public_key.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import os
+import flask
+from loguru import logger
+from config import Config
+
+webapp = flask.Flask(__name__)
+
+
+@webapp.route("/mock/scd/.well-known/uas-traffic-management/mock_pub.der", methods=["GET"])
```
Comment (barroco) [inline monitoring/messagesigning/mockuss_well_known_public_key.py:9]: Is the well-known endpoint service-specific or a general feature USSs will need to enable for all services including Remote ID for instance?
In any case, this implementation will need to move to the [`mock_uss`](https://github.com/interuss/dss/tree/master/monitoring/mock_uss) module, either in a new service folder or in [mock_uss/routes.py](https://github.com/interuss/dss/blob/master/monitoring/mock_uss/routes.py) if this endpoint is unique for a USS.
Please note that the activation of this endpoint should be subject to a configuration flag disabled by default.

File: monitoring/messagesigning/keys/mock_priv.pem [hunk 7]
```diff
@@ -0,0 +1 @@
+Private Key for Mock Uss goes here
```
Comment (barroco) [inline monitoring/messagesigning/keys/mock_priv.pem:1]: A script to generate the keys would be required to allow the end to end testing.

File: monitoring/messagesigning/README.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
+# Message signing reporting/testing set up
```
Comment (barroco) [inline monitoring/messagesigning/README.md:1]: The sections in this file should be move to mock_uss/README.md in a `Message signing` section.

File: monitoring/messagesigning/README.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+There is an environment variable titled `MESSAGE_SIGNING`. It must be set to `"true"` within `run_locally.sh` and `run_locally_scdsc.sh`. This will ensure that outgoing requests are signed and incoming requests, as well as received responses, have their message signing headers validated.
```
Comment (barroco) [inline monitoring/messagesigning/README.md:3]: I don't see the new configuration flag in Config class below. You may want to add it?

File: monitoring/messagesigning/README.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+3. Invalid signatures. This kind of issue means the signature within the `x-utm-message-signature` header was not valid. This could happen for several reasons, it is bessed to check the logs within the `mock_uss` container for more information. 
+
+### Other info
+The `mock_uss` container will sign it's responses. It's important to check that it's signed responses can be validated by your USS. It's also important to test that the outgoing signed requests can be validated by your USS. In addition to this happy path testing, you can test that your USS rejects invalid requests, by turning the environment variable `MESSAGE_SIGNING` to "false". This will disable the signing of outgoing requests, and you should be able to see your USS handles the missing message signing headers request accordingly. It is also possible to leave the `MESSAGE_SIGNING` var to true, but use a faulty private key to sign. USS's should reject these faulty requests with a 403 status code.
```
Comment (barroco) [inline monitoring/messagesigning/README.md:25]: I would recommend to move this section in a new subsection called `Usage` in the new `Message signing` section in `mock_uss/README.md`.

File: monitoring/messagesigning/README.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+### Run well known server
+
+``./run_well_known.sh``  
+Public key will now be accessible at: http://host.docker.internal:8077/mock/scd/.well-known/uas-traffic-management/mock_pub.der
+
+## Creating the message signing report
+1. Inside of `monitoring/uss_qualifier`, run `./run_locally.sh configurations.dev.faa.uft.local_message_signing`
+2. Consult the created report within `monitoring/mock_uss/report/report_mock_uss_scdsc_messagesigning.json`
+
+## Analyzing the report
```
Comment (barroco) [inline monitoring/messagesigning/README.md:18]: This section should not be part of this PR.

File: monitoring/messagesigning/README.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+### Keys
+Keys are under the folder - `monitoring/messagesigning/keys`. This folder should have two keys. Your `uft_priv.pem` key named as `mock_priv.pem`, and your public key used for valdiation named `mock_pub.der`. It is essential these two keys exist in that directory before running any message signing analysis.
+
+### Run well known server
+
+``./run_well_known.sh``  
+Public key will now be accessible at: http://host.docker.internal:8077/mock/scd/.well-known/uas-traffic-management/mock_pub.der
+
+## Creating the message signing report
+1. Inside of `monitoring/uss_qualifier`, run `./run_locally.sh configurations.dev.faa.uft.local_message_signing`
```
Comment (barroco) [inline monitoring/messagesigning/README.md:15]: This section should not be part of this PR.

**Review by barroco — CHANGES_REQUESTED** [review]
Thank you @punamverma for breaking down the PR as recommended by @BenjaminPelletier. Please see my comments on how to better integrate with the current code base, namely, move this new endpoint inside the [mock_uss](https://github.com/interuss/dss/tree/master/monitoring/mock_uss) module and docker image.

File: monitoring/messagesigning/requirements.txt (hunk 4 above)
Comment (kpkeenan12) [reply monitoring/messagesigning/requirements.txt:2]: `loguru` is a nice logging library, we started using it because when calling `print`, we wouldn't see that output. `loguru` started logging our output and we saw it was highly configurable.

File: monitoring/messagesigning/mockuss_well_known_public_key.py (hunk 6 above)
Comment (kpkeenan12) [reply monitoring/messagesigning/mockuss_well_known_public_key.py:9]: mock_uss will sign its responses according to the latest [spec](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-11). This endpoint provides the public key needed to validate these signatures. Other USSs will not necessarily need this endpoint, so long as when they sign their requests/responses, in their `x5u` field, they provide a publicly accessible endpoint that other USSs can use to validate their signatures.

File: monitoring/messagesigning/keys/mock_priv.pem (hunk 7 above)
Comment (kpkeenan12) [reply monitoring/messagesigning/keys/mock_priv.pem:1]: This is tricky. You can gen a pub/priv keypair with `pycryptodome` to use for testing, but these keys will not pass the FAA SCVP validation. Because of this, we asked FAA to give us mock keys that would pass their cert eval, they are located in `monitoring/messagesigning/keys`. This requires the least amount of work since by default these valid keys are going to be used. Based on other comments in this PR, we will add an endpoint to make the keypair to be used configurable, so that negative tests can be run etc, but no need to have to gen a keypair since a valid pair already exists.

File: monitoring/messagesigning/README.md (hunk 9 above)
Comment (kpkeenan12) [reply monitoring/messagesigning/README.md:3]: This is an environment variable as of right now, set similarly to other env vars (`MOCK_USS_DSS_URL`, etc)

Comment (BenjaminPelletier) [reply monitoring/messagesigning/README.md:3]: Note the discussion [here](https://github.com/interuss/dss/pull/899#discussion_r1023171932); it would probably be a good idea to align on design by completing #899 before working on PRs that refer to wider designs not included in the PR (like this one)

File: monitoring/messagesigning/keys/mock_priv.pem (hunk 7 above)
Comment (BenjaminPelletier) [reply monitoring/messagesigning/keys/mock_priv.pem:1]: It would be great to complete #899 to get a good wholistic view of how each component PR will fit into the whole.  For instance, this is a question of what kind of test data will be provided where.  There are a number of different possibilities (including build/test-certs and monitoring/uss_qualifier/test_data) depending on the overall approach.  But, I don't think we've identified a reason why a `monitoring/messagesigning` folder needs to exist -- it seems like it was initially created separately from mock_uss because of [unfamiliarity with mock_uss architecture](https://github.com/interuss/dss/pull/897#discussion_r1022024581).  Using multiple workers and the SynchronizedValue tool used in other parts of mock_uss, this probably doesn't need to be a separate monitoring tool.  It would be great to discuss in #899 (see [architecture suggestion](https://github.com/interuss/dss/pull/899#discussion_r1023154837)).

Comment (kpkeenan12) [reply monitoring/messagesigning/keys/mock_priv.pem:1]: The idea behind making `messagesigning` it's own package in `monitoring` is it won't just be used by `mock_uss`. `mock_uss` uses it to sign responses and validate incoming request http signatures, but it's also used by `monitoring.monitorlib.infrastructure.AuthAdapter` to sign outgoing requests to the USS under test (to ensure that individual USSs are correctly validating request signatures). I will work on the requested changes for `mock_uss` but i still think `messagesigning` should be its own package since it's being shared across components.

Comment (kpkeenan12) [reply monitoring/messagesigning/keys/mock_priv.pem:1]: Linking the [PR](https://github.com/interuss/dss/pull/893) containing the `messagesigning` code

Comment (BenjaminPelletier) [reply monitoring/messagesigning/keys/mock_priv.pem:1]: The best place to put code shared across multiple pieces of monitoring infrastructure is the `monitoring/monitorlib` folder -- either in a single .py file if it's small enough, or, e.g., `monitoring/monitorlib/messagesigning`.  That still leaves the question of where test data should be placed and that would probably depend on what would use the test data and how.

`monitoring.monitorlib.infrastructure.AuthAdapter` (presumably the specific [`SignedRequest` implementation](https://github.com/interuss/dss/blob/37fbff5c070cd899a8e4e1e18ee9ff78def93a22/monitoring/monitorlib/auth.py#L252)) signs outgoing requests to an auth server, but not a USS under test; it would be helpful to clarify expectations, hopefully in #899.

I'm worried about linking to open PRs within a PR.  It's important to have [small and focused PRs](https://github.com/interuss/dss/blob/master/CONTRIBUTING.md#general-principles) and a small and focused PR should not depend on other work in progress.  Linking to another open PR within a PR seems to be suggesting that understanding this PR depends on understanding the linked PR, and that suggests that this PR is not small and focused.  PRs need to be reviewed (and updated, when necessary) and merged one at a time.  I would suggest that #899 is the best candidate for first PR to complete and merge.  #891 has a number of open comments, but updating the dummy auth API YAML does seem like a well-scoped set of changes for a single PR that wouldn't need a design document (like the one in #899).

Comment (kpkeenan12) [reply monitoring/messagesigning/keys/mock_priv.pem:1]: I will focus on #899. To clarify my above comment, I didn't mean `SignedRequest`, I meant the overall parent `AuthAdapter`. To add the required message signing headers, I had to edit the `add_headers` method. This means that all outgoing requests are signed technically, not just to the USS under test.

Comment (kpkeenan12) [reply monitoring/messagesigning/keys/mock_priv.pem:1]: Could you please clarify what you mean by test data? We're reusing test data that already exists. Its basically re-using already existing testing scenarios, and running message signing analysis on the requests/responses associated with that scenario.

**Comment by BenjaminPelletier** [comment]
Closing this PR in favor of #927, though this PR will still be available and could be reopened if appropriate.
